source/UIOverlay.cpp

```diff
@@ -1,21 +1,43 @@
 #include "UIOverlay.hpp"
+#include "Game.hpp"
 
-void UIOverlay::setOverlay(Ogre::Overlay *ov) {
+UIOverlay::UIOverlay(std::string const &name)
+  : overlay(Ogre::OverlayManager::getSingleton().create(name))
+  , buttons()
+{
+  overlay->show();
+}
 
-	overlay = ov;
+UIOverlay::~UIOverlay()
+{
+  overlay->hide();
+  Ogre::OverlayManager::getSingleton().destroy(overlay);
 }
 
 Ogre::Overlay *UIOverlay::getOverlay(void) const {
-
-	return overlay;
+  return overlay;
 }
 
 Ogre::Vector2 UIOverlay::pixelsToRelative(Ogre::Vector2 pixels) {
-
-	return Ogre::Vector2(pixels.x / (Ogre::Real)Game::WIDTH, pixels.y / (Ogre::Real)Game::HEIGHT);
+  return Ogre::Vector2(pixels.x / (Ogre::Real)Game::WIDTH, pixels.y / (Ogre::Real)Game::HEIGHT);
 }
 
 Ogre::Vector2 UIOverlay::relativeToPixels(Ogre::Vector2 relative) {
+  return Ogre::Vector2(relative.x * (Ogre::Real)Game::WIDTH, relative.y * (Ogre::Real)Game::HEIGHT);
+}
 
-	return Ogre::Vector2(relative.x * (Ogre::Real)Game::WIDTH, relative.y * (Ogre::Real)Game::HEIGHT);
+bool UIOverlay::mousePressed(Ogre::Real x, Ogre::Real y)
+{
+  for (auto &&button : buttons) {
+    Ogre::BorderPanelOverlayElement *menuButton(button.second->getPanel());
+    Ogre::Vector2 buttonSize(UIOverlay::relativeToPixels({ menuButton->getWidth(),
+			    menuButton->getHeight() }));
+    Ogre::Vector2 buttonPos(UIOverlay::relativeToPixels({ menuButton->getLeft(),
+			    menuButton->getTop() }));
+    if (x >= buttonPos.x && x <= buttonPos.x + buttonSize.x
+      && y >= buttonPos.y && y <= buttonPos.y + buttonSize.y) {
+      buttons[menuButton->getName()]->getCallback()();
+    }
+  }
```

Comment:
This `for` seems fishy: why the access based on the button's name in `buttons` (the element we are iterating on).
Also usual disclaimer about `operator[]`